0010-zkvm-keccak.md

```diff
@@ -0,0 +1,788 @@
+```
+
+Together with $3\times4=12$ lookups to check that the remainder, quotient and bound are $<2^{64}$.
+
+> Assuming that we execute rotation starting form reset values, then one could perform the above operations directly on the expanded terms, since $4\cdot64-3=253$ encoded bits (removing the three leading zeros) would fit in the BN254 curve. In this scenario, rotations by $x$ bits would be mapped to rotations by an offset of $4\cdot x$ bits instead. Moreover, the constraints would need to be slightly different, to avoid overflows. In particular: `word = quotient * 2^(253-4*offset) + shift` and `output = quotient + shift * 2^(4*offset)`, and the output would already be in expanded format.  Nonetheless, in this case we would also need to perform bound checks, which would be much more costly due to the bitlengths. And in any case, each dense word should be expanded afterwards to be processed in later steps.
+
+### Constraints
+
+
+> For efficiency reasons, the state is assumed to be stored in expanded form, so that back-and-forth conversions do not need to take place repeatedly.
+
+The support for the Keccak hash function will require the following gate types:
+
+**PRECONDITIONS:** the message to be hashed should be padded with the $10^*1$ rule before 
+
+- `KeccakSponge`: performs formatting actions related to the Keccak sponge.
+- `KeccakRound`: performs one full round of the Keccak permutation function.
+
+The high-level layout of the gates follows:
+
+| `KeccakSponge` | [0...100) | [100...168) | [168...200) | [200...400) | [400...800) |
+| -------------- | --------- | ----------- | ----------- | ----------- | ----------- |
+| Curr           | old_state | new_block   | zeros       | bytes      | shifts      |
+| Next           | xor_state |
+
+| `KeccakRound` | [0...440)   | [440...1540) | [1540...2440) | 
+| ------------- | ----------- | ------------ | ------------- | 
+| Curr          | theta_step  | pirho_step   |  chi_step     |
+
+| `KeccakRound` | [0...100)   |
+| ------------- | ----------- |
+| Next          | iota_step   |
+
+#### Sponge
+
+Support for padding shall be provided. In the Keccak PoC, this step took place at the Snarky layer. It checks that the correct amount of bits in the $10^*1$ rule are added until reaching a multiple of 1088 bits, and then adds 512 more zero bits to each block to form a full state.
+
+If the input was not previously expanded, the next step would be to expand all 25 words. Each word of 64 bits would have to be split into 4 parts of 16 real bits each. The expansion itself would be performed through the `Reset` lookup table. This step would require $4\times25=100$ lookups.
+
+There are two main modes: absorb and squeeze. Inside the absorb mode, there are two additional submodes called pad and root, which can as well happen at the same time. 
+
+
+- **Absorb mode**
+
+    In absorb mode, the first coefficient is set to 1: `absorb := coeff[0] = 1`, and the second coefficient is 0.
```

Comment:
I don't understand here: what is coefficient in this case? What is `1` (the one of the field?)

Reply:
coeff refers to the coefficients of the gate. Like in the generic gate, where they are used to encode the operation, here they encode modes of operation. The one is indeed a field element one.